# Отладка тестов через Python Debugger

```bash
# test_pdb.py
def integers_counter(data):
    # Создаем счётчик для целых чисел.
    integers_found = 0
    # Перебираем в цикле элементы входного списка.
    for item in data:
        # Если элемент - целое число, то увеличиваем счётчик.
        if isinstance(item, int):
            integers_found += 1
    # Возвращаем счётчик.
    return integers_found


def test_counter():
    # Произвольные данные для анализа.
    data = [False, 1.0, "some_string", 3, True, 1, [], False]
    # Вызываем функцию:
    integers = integers_counter(data)
    # Целых чисел должно быть 2.
    assert integers == 2
```

- Первым делом надо войти в режим отладки; при работе с pytest это можно сделать несколькими способами:
- - Запустить тесты с ключом `--pdb`. При таком вызове `pdb` будет останавливать работу кода на строках, непосредственно спровоцировавших ошибку. Выполнение файла `test_pdb.py` остановится на строке с assert, но такое поведение дебаггера ничем не поможет в поиске ошибки.
- - Запустить тесты с ключом `--trace`. При таком варианте pdb будет вызываться в начале каждого теста. Запускать в этом режиме успешные тесты не имеет смысла, поэтому лучше комбинировать `--trace` с другими ключами, например `pytest --trace --lf -x`. При такой комбинации ключей в режиме дебаггера будет запущен первый упавший тест, и после окончания теста выполнение будет остановлено.
- - Можно установить точку останова (`breakpoint`) непосредственно в коде теста. Для этого требуется импортировать модуль `pdb` и перед нужной строкой поставить инструкцию: `pdb.set_trace()`. А можно вызвать встроенную функцию `breakpoint()`, которая по умолчанию вызывает pdb.set_trace() (но также может быть переопределена на другой отладчик). При вызове breakpoint() импортировать pdb не нужно.


## PDB (python debaug)

```python
# test_pdb.py
# Импортируйте модуль pdb.
import pdb


def integers_counter(data):
    ...  # Содержимое функции.


def test_counter():
    data = [False, 1.0, "some_string", 3, True, 1, [], False]
    pdb.set_trace()  # Точка останова. Именно отсюда начнём дебаг.
    integers = integers_counter(data)
    assert integers == 2
```

- Выполнение программы остановилось на нужной строке:

```bash
>>>>>>>>>>>>>> PDB set_trace (IO-capturing turned off) >>>>>>>>>>>>>>>>
> <путь до проекта>\test_pdb.py(15)test_counter()
-> integers = integers_counter(data)
(Pdb)
```

В той же строке, где выведен путь до файла, в скобках указан номер строки, на которой сейчас находится отладчик; дальше идёт название функции, которая сейчас выполняется. Дальнейшая работа будет заключаться в пошаговом выполнении программы: это поможет разобраться в происходящем.

- `n` (next) — перейти к следующей строке текущей функции (если в текущей строке вызывается какая-то функция, то pdb не будет «проваливаться» в неё).
- `s` (step) — перейти к следующей строке в текущей или в вызываемой функции.
- `j` (jump) — перейти к указанному номеру строки. По коду можно передвигаться как вперёд, так и назад; есть ряд ограничений, рассмотрим их позже.
- `c` (continue) — продолжить выполнение до следующей точки остановки.
- `p` (print) — напечатать значение переменной или выражения.
- `a` (args) — напечатать аргументы текущей функции.
- `l` (list) — вывести строки кода выполняемой программы.
- `q` (quit) — выйти из режима отладки.
